<a href="https://colab.research.google.com/github/divineachu05/DevOpsDemos/blob/master/Pandas_(9).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1>What is Pandas?</h1>

Pandas is a Python library used for working with data sets.

It has functions for analyzing, cleaning, exploring, and manipulating data.

<b>Why Use Pandas?</b>

Pandas allows us to analyze big data and make conclusions based on statistical theories.

Pandas can clean messy data sets, and make them readable and relevant.

Relevant data is very important in data science.


<b>What Can Pandas Do?</b>

Pandas gives you answers about the data. Like:

Is there a correlation between two or more columns?

What is average value?

Max value?

Min value?

Pandas are also able to delete rows that are not relevant, or contains wrong values, like empty or NULL values. This is called cleaning the data.

Import Pandas

Once Pandas is installed, import it in your applications by adding the import keyword:

In [ ]:
import pandas # pip install pandas if you are getting error

In [ ]:
mydataset = {'cars' :['BMW','Volvo','Ford'],
             'passings':[3,7,2]
    }
myvar = pandas.DataFrame(mydataset)
myvar

cars  passings
0    BMW         3
1  Volvo         7
2   Ford         2

<b>Pandas as pd</b>

Pandas is usually imported under the pd alias.

alias: In Python alias are an alternate name for referring to the same thing.

Create an alias with the as keyword while importing:

In [ ]:
import pandas as pd
mydataset = {'cars' :['BMW','Volvo','Ford'],
             'passings':[3,2,2]
    }
myvar = pd.DataFrame(mydataset)
myvar

cars  passings
0    BMW         3
1  Volvo         2
2   Ford         2

<h1>Pandas Series</h1>

What is a Series?

A Pandas Series is like a column in a table.

It is a one-dimensional array holding data of any type.

Example

Create a simple Pandas Series from a list:

In [ ]:
a = [555,25,89,33,555]
myvar =pd.Series(a)
myvar

0    555
1     25
2     89
3     33
4    555
dtype: int64

<b>Labels</b>

If nothing else is specified, the values are labeled with their index number. First value has index 0, second value has index 1 etc.

This label can be used to access a specified value.

Example

Return the first value of the Series:

In [ ]:
print(myvar[4])

555


<b>Create Labels</b>

With the index argument, you can name your own labels.

Example

Create you own labels:

In [ ]:
a =[1,222,568]
myvar = pd.Series(a,index =['x','y','z'])
myvar

x      1
y    222
z    568
dtype: int64

In [ ]:
print(myvar['z'])

568


<b>Key/Value Objects as Series</b>

You can also use a key/value object, like a dictionary, when creating a Series.

Example

Create a simple Pandas Series from a dictionary:

The keys of the dictionary become the labels.

In [ ]:
cal = {'day1':43,'day2':350,'day3':390}
myvar = pd.Series(cal)
print(myvar)

day1     43
day2    350
day3    390
dtype: int64


To select only some of the items in the dictionary, use the index argument and specify only the items you want to include in the Series.

Example

Create a Series using only data from "day1" and "day2":





In [ ]:
cal = {'day1':43,'day2':350,'day3':390}
myvar = pd.Series(cal,index =['day1','day2'])
print(myvar)

day1     43
day2    350
dtype: int64


<B>What is a DataFrame?</B>

A Pandas DataFrame is a 2 dimensional data structure, like a 2 dimensional array, or a table with rows and columns.

Example
Create a simple Pandas DataFrame:

In [ ]:
data = {'calories':[420,380,390],
       'duration':[50,40,80]}
df =pd.DataFrame(data)
print(df)

   calories  duration
0       420        50
1       380        40
2       390        80


<B>Locate Row</B>

As you can see from the result above, the DataFrame is like a table with rows and columns.

Pandas use the loc attribute to return one or more specified row(s)

Example

Return row 0:

In [ ]:
print(df.loc[[2,0]])

   calories  duration
2       390        80
0       420        50


<b>Named Indexes</b>

With the index argument, you can name your own indexes.

Example
Add a list of names to give each row a name:

In [ ]:
data = {'calories':[420,380,390],
       'duration':[50,40,80]}
df =pd.DataFrame(data,index = ['day1','day2','day3'])
print(df)

      calories  duration
day1       420        50
day2       380        40
day3       390        80


<b>Locate Named Indexes</b>

Use the named index in the loc attribute to return the specified row(s).

Example

Return "day2":

In [ ]:
print(df.loc['day2'])


calories    380
duration     40
Name: day2, dtype: int64


<b>Load Files Into a DataFrame</b>

If your data sets are stored in a file, Pandas can load them into a DataFrame.

Example

Load a comma separated file (CSV file) into a DataFrame:

In [ ]:
df = pd.read_csv('D:\DataScienceBhavana\data.csv')
print(df)

    linear  squared    cubed  power_of_two
0        1        1        1  2.000000e+00
1        2        4        8  4.000000e+00
2        3        9       27  8.000000e+00
3        4       16       64  1.600000e+01
4        5       25      125  3.200000e+01
..     ...      ...      ...           ...
95      96     9216   884736  7.922820e+28
96      97     9409   912673  1.584560e+29
97      98     9604   941192  3.169130e+29
98      99     9801   970299  6.338250e+29
99     100    10000  1000000  1.267650e+30

[100 rows x 4 columns]


<b>Read CSV Files</b>

A simple way to store big data sets is to use CSV files (comma separated files).

CSV files contains plain text and is a well know format that can be read by everyone including Pandas.

In our examples we will be using a CSV file called 'data.csv

Example:
Load the CSV into a DataFrame:

*use to_string() to print the entire DataFrame.

In [ ]:
df = pd.read_csv('D:\DataScienceBhavana\data.csv')
print(df.to_string())

    linear  squared    cubed  power_of_two
0        1        1        1  2.000000e+00
1        2        4        8  4.000000e+00
2        3        9       27  8.000000e+00
3        4       16       64  1.600000e+01
4        5       25      125  3.200000e+01
5        6       36      216  6.400000e+01
6        7       49      343  1.280000e+02
7        8       64      512  2.560000e+02
8        9       81      729  5.120000e+02
9       10      100     1000  1.024000e+03
10      11      121     1331  2.048000e+03
11      12      144     1728  4.096000e+03
12      13      169     2197  8.192000e+03
13      14      196     2744  1.638400e+04
14      15      225     3375  3.276800e+04
15      16      256     4096  6.553600e+04
16      17      289     4913  1.310720e+05
17      18      324     5832  2.621440e+05
18      19      361     6859  5.242880e+05
19      20      400     8000  1.048576e+06
20      21      441     9261  2.097152e+06
21      22      484    10648  4.194304e+06
22      23 

<b>Pandas - Analyzing DataFrames</b>

Viewing the Data

One of the most used method for getting a quick overview of the DataFrame, is the head() method.

The head() method returns the headers and a specified number of rows, starting from the top.

Example
Get a quick overview by printing the first 10 rows of the DataFrame:

In [ ]:
df.head(9)

linear  squared  cubed  power_of_two
0       1        1      1           2.0
1       2        4      8           4.0
2       3        9     27           8.0
3       4       16     64          16.0
4       5       25    125          32.0
5       6       36    216          64.0
6       7       49    343         128.0
7       8       64    512         256.0
8       9       81    729         512.0

There is also a tail() method for viewing the last rows of the DataFrame.

The tail() method returns the headers and a specified number of rows, starting from the bottom.

Example

Print the last 5 rows of the DataFrame:

In [ ]:
df.tail(10)

linear  squared    cubed  power_of_two
90      91     8281   753571  2.475880e+27
91      92     8464   778688  4.951760e+27
92      93     8649   804357  9.903520e+27
93      94     8836   830584  1.980700e+28
94      95     9025   857375  3.961410e+28
95      96     9216   884736  7.922820e+28
96      97     9409   912673  1.584560e+29
97      98     9604   941192  3.169130e+29
98      99     9801   970299  6.338250e+29
99     100    10000  1000000  1.267650e+30

<b>Info About the Data</b>

The DataFrames object has a method called info(), that gives you more information about the data set.

Example:
Print information about the data:

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   linear        100 non-null    int64  
 1   squared       100 non-null    int64  
 2   cubed         100 non-null    int64  
 3   power_of_two  100 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 3.2 KB


<b>Null Values</b>

The info() method also tells us how many Non-Null values there are present in each column, and in our data set it seems like there are 164 of 169 Non-Null values in the "Calories" column.

Which means that there are 5 rows with no value at all, in the "Calories" column, for whatever reason.

Empty values, or Null values, can be bad when analyzing data, and you should consider removing rows with empty values. This is a step towards what is called cleaning data.

<b>Data Cleaning</b>

Data cleaning means fixing bad data in your data set.

Bad data could be:

Empty cells

Data in wrong format

Wrong data

Duplicates

Empty cells can potentially give you a wrong result when you analyze data.

Remove Rows :
One way to deal with empty cells is to remove rows that contain empty cells.

This is usually OK, since data sets can be very big, and removing a few rows will not have a big impact on the result.

Example:
Return a new Data Frame with no empty cells:

In [ ]:
df = pd.read_csv('D:\DataScienceBhavana\dirtydata.csv')
new_df = df.dropna()
print(new_df)

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'  

Note: By default, the dropna() method returns a new DataFrame, and will not change the original.

If you want to change the original DataFrame, use the inplace = True argument:

Example: 
Remove all rows with NULL values:

In [ ]:
df = pd.read_csv('D:\DataScienceBhavana\dirtydata.csv')
df.dropna(inplace = True)
print(df)

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'  

<b>Replace Empty Values</b>

Another way of dealing with empty cells is to insert a new value instead.

This way you do not have to delete entire rows just because of some empty cells.

The fillna() method allows us to replace empty cells with a value:

Example:
Replace NULL values with the number 130:

In [ ]:
df = pd.read_csv('D:\DataScienceBhavana\dirtydata.csv')
df.fillna(150,inplace =True)
df

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112     150.0
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           150    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132     150.0
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

<b>Replace Only For a Specified Columns</b>

The example above replaces all empty cells in the whole Data Frame.

To only replace empty values for one column, specify the column name for the DataFrame:

Example:
Replace NULL values in the "Calories" columns with the number 130:

In [ ]:
df = pd.read_csv('D:\DataScienceBhavana\dirtydata.csv')
df['Calories'].fillna(130,inplace=True)
print(df)

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'  

<b>Replace Using Mean, Median, or Mode</b>

A common way to replace empty cells, is to calculate the mean, median or mode value of the column.

Pandas uses the mean() median() and mode() methods to calculate the respective values for a specified column:

Example:Calculate the MEAN, and replace any empty values with it:

In [ ]:
df = pd.read_csv('D:\DataScienceBhavana\dirtydata.csv')

x =df['Calories'].mean()
print(x)
df['Calories'].fillna(x,inplace = True)
df

304.68


Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7        450  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
12        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120    215.20
17        60  '2020/12/17'    100       120    300.00
18        45  '2020/12/18'     90       112    304.68
19        60  '2020/12/19'    103       123    323.00
20        45  '2020/12/20'     97       125    243.00
21        60  '2020/12/21'    108       131    364.20
22        45           NaN    100       119    282.00
23        60  '2020/12/23'    130       101    300.00
24        45  '2020/12/24'    105       132    246.00
25        60  '2020/12/25'    102       126    334.50
26        60      20201226    100       120    250.00
27        60  '2020/12/27'     92       118    241.00
28        60  '2020/12/28'    103       132    304.68
29        60  '2020/12/29'    100       132    280.00
30        60  '2020/12/30'    102       129    380.30
31        60  '2020/12/31'     92       115    243.00

Mean = the average value (the sum of all values divided by number of values).

Example:
Calculate the MEDIAN, and replace any empty values with it:

In [ ]:
df = pd.read_csv('D:\DataScienceBhavana\dirtydata.csv')

x =df['Calories'].median()
print(x)
df['Calories'].fillna(x,inplace = True)
df

291.2


Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112     291.2
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132     291.2
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

Median = the value in the middle, after you have sorted all values ascending.

Example
Calculate the MODE, and replace any empty values with it:

In [ ]:
import pandas as pd
df = pd.read_csv('D:\DataScienceBhavana\dirtydata.csv')

x =df['Calories'].mode()[0]
print(x)
df['Calories'].fillna(x,inplace = True)
df

300.0


Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112     300.0
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132     300.0
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

<h1>Pandas - Cleaning Data of Wrong Format</h1>

Data of Wrong Format:
    
Cells with data of wrong format can make it difficult, or even impossible, to analyze data.

To fix it, you have two options: remove the rows, or convert all cells in the columns into the same format.

Convert Into a Correct Format
In our Data Frame, we have two cells with the wrong format. Check out row 22 and 26, the 'Date' column should be a string that represents a date:

Let's try to convert all cells in the 'Date' column into dates.

Pandas has a to_datetime() method for this:

Example
Convert to date:



In [ ]:
import pandas as pd
df = pd.read_csv("D:\DataScienceBhavana\dirtydata.csv")
df['Date'] = pd.to_datetime(df['Date'])
print(df.to_string())

    Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90 

As you can see from the result, the date in row 26 where fixed, but the empty date in row 22 got a NaT (Not a Time) value, in other words an empty value. One way to deal with empty values is simply removing the entire row.

Removing Rows

The result from the converting in the example above gave us a NaT value, which can be handled as a NULL value, and we can remove the row by using the dropna() method.

Example
Remove rows with a NULL value in the "Date" column:

In [ ]:
df['Date'].dropna(inplace = True)
df

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

<b>Wrong Data</b>

"Wrong data" does not have to be "empty cells" or "wrong format", it can just be wrong, like if someone registered "199" instead of "1.99".

Sometimes you can spot wrong data by looking at the data set, because you have an expectation of what it should be.

If you take a look at our data set, you can see that in row 7, the duration is 450, but for all the other rows the duration is between 30 and 60.

It doesn't have to be wrong, but taking in consideration that this is the data set of someone's workout sessions, we conclude with the fact that this person did not work out in 450 minutes.


How can we fix wrong values, like the one for "Duration" in row 7?

<b>Replacing Values:</b>

One way to fix wrong values is to replace them with something else.

In our example, it is most likely a typo, and the value should be "45" instead of "450", and we could just insert "45" in row 7:

Example:
Set "Duration" = 45 in row 7:

In [ ]:
df.loc[7,'Duration'] = 45
df

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7         45 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

For small data sets you might be able to replace the wrong data one by one, but not for big data sets.

To replace wrong data for larger data sets you can create some rules, e.g. set some boundaries for legal values, and replace any values that are outside of the boundaries.

Example:
Loop through all values in the "Duration" column.

If the value is higher than 120, set it to 120:

In [ ]:
for x in df.index:
    if df.loc[x,'Duration']>120:
        df.loc[x,"Duration"] = 120

<b>Removing Rows</b>
Another way of handling wrong data is to remove the rows that contains wrong data.

This way you do not have to find out what to replace them with, and there is a good chance you do not need them to do your analyses.

Example
Delete rows where "Duration" is higher than 120:

In [ ]:
for x in df.index:
    if df.loc[x,'Duration']>120:
        df.drop(x,inplace =True)

<b>Discovering Duplicates</b>

Duplicate rows are rows that have been registered more than one time.

By taking a look at our test data set, we can assume that row 11 and 12 are duplicates.

To discover duplicates, we can use the duplicated() method.

The duplicated() method returns a Boolean values for each row:

Example
Returns True for every row that is a duplicate, othwerwise False:

In [ ]:
df

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7         45 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

In [ ]:
print(df.duplicated())

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12     True
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
dtype: bool


Removing Duplicates:
To remove duplicates, use the drop_duplicates() method.

Example:
Remove all duplicates:

The (inplace = True) will make sure that the method does NOT return a new DataFrame, but it will remove all duplicates from the original DataFrame.

In [ ]:
df.drop_duplicates(inplace = True)

In [ ]:
df['Duration'].unique()

array([60, 45, 30], dtype=int64)

<h2>Pandas - Data Correlations</h2>

Finding Relationships

A great aspect of the Pandas module is the corr() method.

The corr() method calculates the relationship between each column in your data set.

Example
Show the relationship between the columns:

In [ ]:
df = pd.read_csv("D:\DataScienceBhavana\dirtydata.csv")
df.corr()

Duration     Pulse  Maxpulse  Calories
Duration  1.000000  0.004410  0.049959 -0.114169
Pulse     0.004410  1.000000  0.276583  0.513186
Maxpulse  0.049959  0.276583  1.000000  0.357460
Calories -0.114169  0.513186  0.357460  1.000000


The Result of the corr() method is a table with a lot of numbers that represents how well the relationship is between two columns.

The number varies from -1 to 1.

1 means that there is a 1 to 1 relationship (a perfect correlation), and for this data set, each time a value went up in the first column, the other one went up as well.

0.9 is also a good relationship, and if you increase one value, the other will probably increase as well.

-0.9 would be just as good relationship as 0.9, but if you increase one value, the other will probably go down.

0.2 means NOT a good relationship, meaning that if one value goes up does not mean that the other will.

What is a good correlation? It depends on the use, but I think it is safe to say you have to have at least 0.6 (or -0.6) to call it a good correlation.

Perfect Correlation:
We can see that "Duration" and "Duration" got the number 1.000000, which makes sense, each column always has a perfect relationship with itself.

Good Correlation:
"Duration" and "Calories" got a 0.922721 correlation, which is a very good correlation, and we can predict that the longer you work out, the more calories you burn, and the other way around: if you burned a lot of calories, you probably had a long work out.

Bad Correlation:
"Duration" and "Maxpulse" got a 0.009403 correlation, which is a very bad correlation, meaning that we can not predict the max pulse by just looking at the duration of the work out, and vice versa.



DataFrame Basic Functionality
Let us now understand what DataFrame Basic Functionality is. The following tables lists down the important attributes or methods that help in DataFrame Basic Functionality.

T : Transposes rows and columns.

axes : Returns a list with the row axis labels and column axis labels as the only members.


dtypes : Returns the dtypes in this object.

empty : True if NDFrame is entirely empty [no items]; if any of the axes are of length 0.
	
ndim : Number of axes / array dimensions.

shape : Returns a tuple representing the dimensionality of the DataFrame.

size : Number of elements in the NDFrame.

values : Numpy representation of NDFrame.

head() : Returns the first n rows.

tail() : Returns last n rows.

Let us now create a DataFrame and see all how the above mentioned attributes operate.

Example:

In [ ]:
import numpy as np
d = {"Name" : pd.Series(['Jack','Ricky','Vin','Steve','Smith','Tom','James']),
    'Age':pd.Series([25,26,27,28,30,33,65]),
    'Rating' : pd.Series([4.2,6.2,8.5,9.3,5.5,3.3,2.2])}

df = pd.DataFrame(d)
print(df)

    Name  Age  Rating
0   Jack   25     4.2
1  Ricky   26     6.2
2    Vin   27     8.5
3  Steve   28     9.3
4  Smith   30     5.5
5    Tom   33     3.3
6  James   65     2.2


<b>T (Transpose)</b>

Returns the transpose of the DataFrame. The rows and columns will interchange.

In [ ]:
df.T

0      1    2      3      4    5      6
Name    Jack  Ricky  Vin  Steve  Smith  Tom  James
Age       25     26   27     28     30   33     65
Rating   4.2    6.2  8.5    9.3    5.5  3.3    2.2

<b>axes</b>

Returns the list of row axis labels and column axis labels.

In [ ]:
df.axes

[RangeIndex(start=0, stop=7, step=1),
 Index(['Name', 'Age', 'Rating'], dtype='object')]

<b>dtypes</b>
Returns the data type of each column.

In [ ]:
df.dtypes

Name       object
Age         int64
Rating    float64
dtype: object

<b>empty</b>
Returns the Boolean value saying whether the Object is empty or not; True indicates that the object is empty.

In [ ]:
df.empty

False

<b>ndim</b>
Returns the number of dimensions of the object. By definition, DataFrame is a 2D object.

In [ ]:
df.ndim

2

In [ ]:
df

Name  Age  Rating
0   Jack   25     4.2
1  Ricky   26     6.2
2    Vin   27     8.5
3  Steve   28     9.3
4  Smith   30     5.5
5    Tom   33     3.3
6  James   65     2.2

<b>shape</b>
Returns a tuple representing the dimensionality of the DataFrame. Tuple (a,b), where a represents the number of rows and b represents the number of columns.

In [ ]:
df.shape

(7, 3)

<b>size</b>
Returns the number of elements in the DataFrame.

In [ ]:
df.size

21

<b>values</b>
Returns the actual data in the DataFrame as an NDarray.

In [ ]:
df.values

array([['Jack', 25, 4.2],
       ['Ricky', 26, 6.2],
       ['Vin', 27, 8.5],
       ['Steve', 28, 9.3],
       ['Smith', 30, 5.5],
       ['Tom', 33, 3.3],
       ['James', 65, 2.2]], dtype=object)

<b>The describe() Function^</b>

It returns all the necessary values such as the count, min, max, mean, median etc. for a series.

In [ ]:
df.describe()

Age    Rating
count   7.000000  7.000000
mean   33.428571  5.600000
std    14.175767  2.622975
min    25.000000  2.200000
25%    26.500000  3.750000
50%    28.000000  5.500000
75%    31.500000  7.350000
max    65.000000  9.300000

As you can see, we have got the summary of descriptive statistics for the `df` series. You can also spot some of the values against the `25%, 50%` and `75%` parameters. They are called percentiles (or quartiles). Let's learn each of them one-by-one. Let's start with `25%`.

**Percentiles (or Quartiles)**

`25%` refers to the 
$25^{th}$ 
percentile value. It is also called the **first quartile** value. *The first quartile is defined as the middle number between the smallest number and the median of the dataset.* 

Let's understand this concept through an example. Consider the following series of numbers,

$3, 9, 6, 10, 7, 9, 7, 6, 1$

After arranging them in the increasing order, we have the following series

$1, 3, 6, 6, 7, 7, 9, 9, 10$

The median value of the series is 
$7$ 
because it is the middle value. So, the 
$25^{th}$ 
percentile (or first quartile) value will be the median of the following values

$1, 3, 6, 6, 7$

Therefore, first quartile or 
$Q1 = 6$ 
because 
$6$ 
is the median of the above sub-series.

**Note:** The median value is also referred to as the **second quartile** (or 
$50^{th}$ 
percentile) value.

Similarly, `75%` refers to the 
$75^{th}$ 
percentile value. It is also called the **third quartile** value. *The third quartile is the middle value between the median and the highest value of the dataset.*

So, the 
$75^{th}$ 
percentile (or third quartile) value will be the median of the following values

$7, 7, 9, 9, 10$

Therefore, third quartile or 
$Q3 = 9$ 
because 
$9$ 
is the median of the above sub-series.

*The first, second and third quartile values together tell the spread of the data. If a majority of the data points are spread apart from each other, then there will be a significantly large difference between the three quartile values whereas if a majority of the data points are close to each other, then all three quartile values will also be close to each other.*

In the 

$3, 9, 6, 10, 7, 9, 7, 6, 1$

series, the first, second and third quartile values are 
$Q1=6, Q2=7$ 
and 
$Q3=9$ 
respectively. This means the majority of the data points are quite close to each other which is quite evident from the series.


**Standard Deviation (or `std`)**

*The standard deviation is the square root of the mean of squared deviations.* It tells us how far away a data point is from the mean value. Larger the standard deviation, farther the data point from the mean value.

For the above series, the mean value
$(\mu)$
is 

$\mu = \frac{3+9+6+10+7+9+7+6+1}{9} = 6.44$

The **deviation** is the difference between the mean value and the individual data points in a series, i.e., 

$(3 - 6.44), (9 - 6.44), (6 - 6.44), (10 - 6.44), (7 - 6.44), (9 - 6.44), (7 - 6.44), (6 - 6.44), (1 - 6.44)$

are the deviations in the series. 

The **squared deviation** is the square of deviations, i.e., 

$(3 - 6.44)^2, (9 - 6.44)^2, (6 - 6.44)^2, (10 - 6.44)^2, (7 - 6.44)^2, (9 - 6.44)^2, (7 - 6.44)^2, (6 - 6.44)^2, (1 - 6.44)^2$

are the squared deviations in the series.

The **mean of squared deviations (or MSD)** is a sum of all the squared deviations divided by the number number of values, i.e., 

$\frac{(3 - 6.44)^2 + (9 - 6.44)^2 + (6 - 6.44)^2 + (10 - 6.44)^2 + (7 - 6.44)^2 + (9 - 6.44)^2 + (7 - 6.44)^2 + (6 - 6.44)^2 + (1 - 6.44)^2}{9}$

So, the standard deviation (or 
$std$
) is

$\sqrt{\frac{(3 - 6.44)^2 + (9 - 6.44)^2 + (6 - 6.44)^2 + (10 - 6.44)^2 + (7 - 6.44)^2 + (9 - 6.44)^2 + (7 - 6.44)^2 + (6 - 6.44)^2 + (1 - 6.44)^2}{9}}$

$\Rightarrow std = 2.92$

The significance of the standard deviation is that the numbers, 
$9, 6, 7, 9, 7, 6$
, from the series, lie in the range 
$(mean - std)$
and
$(mean + std)$
, i.e., 
$6.44 - 2.92 = 3.52$
and
$6.44 + 2.92 = 9.36$

So, these numbers are closest to the mean value.

The remaining three numbers, 
$1, 3$
and
$10$
lie in the range
$(mean - 2\times std)$
and
$(mean + 2\times std)$
. So, these three numbers are farthest from the mean value.

We can verify the above calculations by first creating a Pandas series using the `pd.Series()` function and then by applying the `describe()` function on the 
$3, 9, 6, 10, 7, 9, 7, 6, 1$ 
series.

**Note:** The MSD is also called **variance**. Therefore, 
$variance = \sigma^2$
where 
$\sigma$ 
is the standard deviation.

In [ ]:
l = ['f','e','d','c','b','a']
a = l[::-1]
type(a)


list

In [ ]:
a = [1,2,3]
b = a.reverse()
b


In [ ]:
type(b)

NoneType

#### Activity 1: Univariate Grouping

Before we begin grouping and aggregation, let's first get the list of columns we are left with after applying all the previous operations.

In [ ]:
#Grouping and Aggregations
# Run the code cell.
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')


# Loading the dataset.
csv_file = 'https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/air-quality/AirQualityUCI.csv'
df = pd.read_csv(csv_file, sep=';')

# Dropping the 'Unnamed: 15' & 'Unnamed: 16' columns.
df = df.drop(columns=['Unnamed: 15', 'Unnamed: 16'], axis=1) 

# Dropping the null values.
df = df.dropna()

# Creating a Pandas series containing 'datetime' objects.
dt_series = df['Date'] + ' ' + pd.Series(data=[str(item).replace(".", ":") for item in df['Time']], index=df.index)
dt_series = pd.to_datetime(dt_series)

# Remove the Date & Time columns from the DataFrame and insert the 'dt_series' in it.
df = df.drop(columns=['Date', 'Time'], axis=1)
df.insert(loc=0, column='DateTime', value=dt_series)

# Get the Pandas series containing the year values as integers.
year_series = dt_series.dt.year

# Get the Pandas series containing the month values as integers.
month_series = dt_series.dt.month

# Get the Pandas series containing the day values as integers.
day_series = dt_series.dt.day

# Get the Pandas series containing the days of a week, i.e., Monday, Tuesday, Wednesday etc.
day_name_series = dt_series.dt.day_name()

# Add the 'Year', 'Month', 'Day' and 'Day Name' columns to the DataFrame.
df['Year'] = year_series
df['Month'] = month_series
df['Day'] = day_series
df['Day Name'] = day_name_series

# Sort the DataFrame by the 'DateTime' values in the ascending order. Also, display the first 10 rows of the DataFrame.
df = df.sort_values(by='DateTime')

# Create a function to replace the commas with periods in a Pandas series.
def comma_to_period(series):
    new_series = pd.Series(data=[float(str(item).replace(',', '.')) for item in series], index=df.index)
    return new_series

# Apply the 'comma_to_period()' function on the ''CO(GT)', 'C6H6(GT)', 'T', 'RH' and 'AH' columns.
cols_to_correct = ['CO(GT)', 'C6H6(GT)', 'T', 'RH', 'AH'] # Create a list of column names.
for col in cols_to_correct: # Iterate through each column
    df[col] = comma_to_period(df[col]) # Replace the original column with the new series.

# Remove all the columns from the 'df' DataFrame containing more than 10% garbage value.
df = df.drop(columns=['NMHC(GT)', 'CO(GT)', 'NOx(GT)', 'NO2(GT)'], axis=1)

# Create a new DataFrame containing records for the years 2004 and 2005.
aq_2004_df = df[df['Year'] == 2004]
aq_2005_df = df[df['Year'] == 2005]

# Replace the -200 value with the median values for each column having indices between 1 and -4 (excluding -4) for the 2004 year DataFrame.
for col in aq_2004_df.columns[1:-4]:
  median = aq_2004_df.loc[aq_2004_df[col] != -200, col].median() # Get the median value for each column after excluding -200.
  aq_2004_df[col] = aq_2004_df[col].replace(to_replace=-200, value=median)

# Repeat the same exercise for the 2005 year DataFrame.
for col in aq_2005_df.columns[1:-4]:
  median = aq_2005_df.loc[aq_2005_df[col] != -200, col].median()
  aq_2005_df[col] = aq_2005_df[col].replace(to_replace=-200, value=median)


In [ ]:
df.columns

Index(['DateTime', 'PT08.S1(CO)', 'C6H6(GT)', 'PT08.S2(NMHC)', 'PT08.S3(NOx)',
       'PT08.S4(NO2)', 'PT08.S5(O3)', 'T', 'RH', 'AH', 'Year', 'Month', 'Day',
       'Day Name'],
      dtype='object')

In [ ]:
df.head()

DateTime  PT08.S1(CO)  C6H6(GT)  PT08.S2(NMHC)  PT08.S3(NOx)  \
0 2004-03-10 18:00:00       1360.0      11.9         1046.0        1056.0   
1 2004-03-10 19:00:00       1292.0       9.4          955.0        1174.0   
2 2004-03-10 20:00:00       1402.0       9.0          939.0        1140.0   
3 2004-03-10 21:00:00       1376.0       9.2          948.0        1092.0   
4 2004-03-10 22:00:00       1272.0       6.5          836.0        1205.0   

   PT08.S4(NO2)  PT08.S5(O3)     T    RH      AH  Year  Month  Day   Day Name  
0        1692.0       1268.0  13.6  48.9  0.7578  2004      3   10  Wednesday  
1        1559.0        972.0  13.3  47.7  0.7255  2004      3   10  Wednesday  
2        1555.0       1074.0  11.9  54.0  0.7502  2004      3   10  Wednesday  
3        1584.0       1203.0  11.0  60.0  0.7867  2004      3   10  Wednesday  
4        1490.0       1110.0  11.2  59.6  0.7888  2004      3   10  Wednesday

The description for all the columns containing data for air pollutants, temperature, relative humidity and absolute humidity is provided below.


|Columns|Description|
|-|-|
|PT08.S1(CO)|PT08.S1 (tin oxide) hourly averaged sensor response (nominally $\text{CO}$ targeted)|
|C6H6(GT)|True hourly averaged Benzene concentration in $\frac{\mu g}{m^3}$|
|PT08.S2(NMHC)|PT08.S2 (titania) hourly averaged sensor response (nominally $\text{NMHC}$ targeted)|
|PT08.S3(NOx)|PT08.S3 (tungsten oxide) hourly averaged sensor response (nominally $\text{NO}_x$ targeted)|
|PT08.S4(NO2)|PT08.S4 (tungsten oxide) hourly averaged sensor response (nominally $\text{NO}_2$ targeted)|
|PT08.S5(O3) |PT08.S5 (indium oxide) hourly averaged sensor response (nominally $\text{O}_3$ targeted)|
|T|Temperature in °C|
|RH|Relative Humidity (%)|
|AH|AH Absolute Humidity|

**The `groupby()` Function**

Now, let's group all the values by months, i.e., club all the values together for January, February and so on...

To group the values of a DataFrame by a particular value, use the `groupby()` function. Inside the function, you have to pass the column name to `by` parameter to specify the column about which you want to group the values.

**Syntax:** `dataframe.groupby(by='column')`

You can additionally set the `sort` parameter either equal to `True` or `False` depending on whether you want to sort the values or not. By default, the `groupby()` function returns a new Pandas `DataFrameGroupBy` object containing the grouped sorted values. 

In [ ]:
group_month_2004 = aq_2004_df.groupby(by = 'Month')
group_month_2004

The `group_by()` function returns a `DataFrameGroupBy` object with its physical memory location.

How you want to group the values depends on the problem statement at hand. Here, we would like to analyse data month-wise. Hence, we have grouped the DataFrame by the `Month` column.

**The `get_group()` Function**

To get all the occurrences of a value from a group, use the `get_group()` function. 

**Syntax:** `DataFrameGroupBy_object.get_group(item)`

E.g., if you want to get all the records for the month of March from the `group_2004_month DataFrameGroupBy` object, pass `3` as input to the `get_group()` function.

In [ ]:
group_month_2004.get_group(4)


DateTime  PT08.S1(CO)  C6H6(GT)  PT08.S2(NMHC)  PT08.S3(NOx)  \
510  2004-04-01 00:00:00       1143.0       6.3          825.0         986.0   
511  2004-04-01 01:00:00       1044.0       5.1          770.0        1031.0   
512  2004-04-01 02:00:00       1034.0       4.1          716.0        1085.0   
513  2004-04-01 03:00:00        956.0       4.0          713.0        1099.0   
514  2004-04-01 04:00:00        909.0       2.4          615.0        1237.0   
...                  ...          ...       ...            ...           ...   
1225 2004-04-30 19:00:00       1626.0      26.9         1476.0         555.0   
1226 2004-04-30 20:00:00       1449.0      19.5         1282.0         625.0   
1227 2004-04-30 21:00:00       1363.0      15.1         1152.0         684.0   
1228 2004-04-30 22:00:00       1371.0      14.6         1136.0         689.0   
1229 2004-04-30 23:00:00       1406.0      13.7         1107.0         718.0   

      PT08.S4(NO2)  PT08.S5(O3)     T    RH      AH  Year  Month  Day  \
510         1477.0        978.0  12.0  61.6  0.8593  2004      4    1   
511         1425.0        944.0  11.5  63.9  0.8652  2004      4    1   
512         1405.0        891.0  10.7  67.2  0.8630  2004      4    1   
513         1422.0        849.0   9.0  73.1  0.8394  2004      4    1   
514         1322.0        790.0  10.2  66.6  0.8299  2004      4    1   
...            ...          ...   ...   ...     ...   ...    ...  ...   
1225        2356.0       1786.0  19.8  58.6  1.3392  2004      4   30   
1226        2100.0       1569.0  19.1  61.1  1.3345  2004      4   30   
1227        1951.0       1495.0  18.2  65.4  1.3529  2004      4   30   
1228        1927.0       1471.0  18.1  66.1  1.3579  2004      4   30   
1229        1872.0       1384.0  17.7  66.9  1.3422  2004      4   30   

      Day Name  
510   Thursday  
511   Thursday  
512   Thursday  
513   Thursday  
514   Thursday  
...        ...  
1225    Friday  
1226    Friday  
1227    Friday  
1228    Friday  
1229    Friday  

[720 rows x 14 columns]

The above DataFrame contains all the records only for April 2004.

Similarly, you can get values for all other months from the `grouped_2004_month DataFrameGroupBy` object.

Now, you group the records of 2005 about the `Month` column so that we can later compare the situation in 2004 with 2005.

In [ ]:
group_month_2005 = aq_2005_df.groupby(by = 'Month')
group_month_2005

#### Activity 2: Aggregation on Univariate Grouping^

You have already learnt the aggregation functions such as `count(), sum(), min(), max(), mean()` etc. Let's apply a few of them on the `DataFrameGroupBy` objects that we have already created.

The pollution is most severe during the winter season in any part of the world. The winter season in Italy begins from the last week of October and lasts till the last week of March. So, let's get the descriptive statistics for the concentrations of the air pollutants, temperature, relative humidity and absolute humidity in March, November and December in 2004 and 2005.

In [ ]:
group_month_2004.get_group(3).describe()

PT08.S1(CO)    C6H6(GT)  PT08.S2(NMHC)  PT08.S3(NOx)  PT08.S4(NO2)  \
count   510.000000  510.000000     510.000000    510.000000    510.000000   
mean   1222.819608    9.932745     935.643137   1029.172549   1572.072549   
std     241.038541    7.060527     258.997406    268.130898    280.525357   
min     818.000000    0.600000     457.000000    537.000000   1050.000000   
25%    1027.250000    4.600000     742.250000    834.250000   1359.000000   
50%    1197.000000    8.650000     926.000000    966.500000   1543.000000   
75%    1383.000000   13.650000    1106.250000   1175.500000   1730.000000   
max    2040.000000   39.200000    1754.000000   1935.000000   2679.000000   

       PT08.S5(O3)           T          RH          AH    Year  Month  \
count   510.000000  510.000000  510.000000  510.000000   510.0  510.0   
mean   1027.494118   14.390784   50.170784    0.789413  2004.0    3.0   
std     381.600387    4.447539   14.265312    0.139932     0.0    0.0   
min     341.000000    6.100000   14.900000    0.402300  2004.0    3.0   
25%     762.500000   11.000000   38.550000    0.691175  2004.0    3.0   
50%     992.500000   14.050000   52.200000    0.792050  2004.0    3.0   
75%    1272.000000   16.800000   60.750000    0.895900  2004.0    3.0   
max    2359.000000   29.300000   83.200000    1.094500  2004.0    3.0   

              Day  
count  510.000000  
mean    20.870588  
std      6.141338  
min     10.000000  
25%     16.000000  
50%     21.000000  
75%     26.000000  
max     31.000000

In [ ]:
group_month_2005.get_group(3).describe()

PT08.S1(CO)    C6H6(GT)  PT08.S2(NMHC)  PT08.S3(NOx)  PT08.S4(NO2)  \
count   744.000000  744.000000     744.000000    744.000000    744.000000   
mean   1145.223118    8.726478     887.740591    709.447581   1225.188172   
std     197.140374    6.677222     253.315169    216.393926    275.936945   
min     715.000000    0.200000     387.000000    330.000000    551.000000   
25%     996.000000    3.700000     692.500000    551.750000   1054.750000   
50%    1126.000000    6.600000     841.000000    684.500000   1217.000000   
75%    1267.500000   12.100000    1052.500000    830.000000   1376.000000   
max    1818.000000   35.500000    1675.000000   1804.000000   2147.000000   

       PT08.S5(O3)           T          RH          AH    Year  Month  \
count   744.000000  744.000000  744.000000  744.000000   744.0  744.0   
mean   1083.598118   13.213306   50.909677    0.781531  2005.0    3.0   
std     408.951492    6.394258   16.598389    0.291686     0.0    0.0   
min     221.000000   -1.900000   13.500000    0.184700  2005.0    3.0   
25%     808.500000    8.675000   37.400000    0.591475  2005.0    3.0   
50%    1039.500000   13.500000   51.800000    0.783400  2005.0    3.0   
75%    1365.250000   17.600000   64.575000    0.991450  2005.0    3.0   
max    2159.000000   28.800000   84.000000    1.393000  2005.0    3.0   

              Day  
count  744.000000  
mean    16.000000  
std      8.950289  
min      1.000000  
25%      8.000000  
50%     16.000000  
75%     24.000000  
max     31.000000

If we keep on aggregating the values month-wise one-by-one, then it will become a cumbersome task. Let's get the descriptive statistics for all the months for both the years 2004 and 2005.

In [ ]:
group_month_2004.describe()

PT08.S1(CO)                                                            \
            count         mean         std    min      25%     50%      75%   
Month                                                                         
3           510.0  1222.819608  241.038541  818.0  1027.25  1197.0  1383.00   
4           720.0  1158.993056  227.826933  753.0   985.75  1115.5  1317.50   
5           744.0  1076.350806  193.766317  736.0   937.75  1053.0  1196.00   
6           720.0  1022.862500  157.922850  708.0   892.75  1030.5  1121.50   
7           744.0  1046.452957  158.314118  723.0   930.00  1032.0  1138.50   
8           744.0   986.259409  117.964334  711.0   900.75   987.5  1060.00   
9           720.0  1081.022222  197.420922  718.0   926.75  1060.0  1217.25   
10          744.0  1184.745968  243.017236  704.0   984.00  1159.5  1356.25   
11          720.0  1132.148611  273.058383  647.0   912.75  1106.5  1311.00   
12          744.0  1084.196237  211.089768  691.0   927.75  1060.0  1199.00   

              C6H6(GT)             ...    Year            Day             \
          max    count       mean  ...     75%     max  count       mean   
Month                              ...                                     
3      2040.0    510.0   9.932745  ...  2004.0  2004.0  510.0  20.870588   
4      1875.0    720.0  10.329583  ...  2004.0  2004.0  720.0  15.500000   
5      1763.0    744.0  10.192339  ...  2004.0  2004.0  744.0  16.000000   
6      1506.0    720.0  10.500972  ...  2004.0  2004.0  720.0  15.500000   
7      1626.0    744.0  10.597043  ...  2004.0  2004.0  744.0  16.000000   
8      1393.0    744.0   7.112634  ...  2004.0  2004.0  744.0  16.000000   
9      1728.0    720.0  11.485694  ...  2004.0  2004.0  720.0  15.500000   
10     1908.0    744.0  13.518952  ...  2004.0  2004.0  744.0  16.000000   
11     2008.0    720.0  12.511806  ...  2004.0  2004.0  720.0  15.500000   
12     1881.0    744.0   9.423656  ...  2004.0  2004.0  744.0  16.000000   

                                               
            std   min   25%   50%   75%   max  
Month                                          
3      6.141338  10.0  16.0  21.0  26.0  31.0  
4      8.661458   1.0   8.0  15.5  23.0  30.0  
5      8.950289   1.0   8.0  16.0  24.0  31.0  
6      8.661458   1.0   8.0  15.5  23.0  30.0  
7      8.950289   1.0   8.0  16.0  24.0  31.0  
8      8.950289   1.0   8.0  16.0  24.0  31.0  
9      8.661458   1.0   8.0  15.5  23.0  30.0  
10     8.950289   1.0   8.0  16.0  24.0  31.0  
11     8.661458   1.0   8.0  15.5  23.0  30.0  
12     8.950289   1.0   8.0  16.0  24.0  31.0  

[10 rows x 88 columns]

We still have quite large DataFrames containing the descriptive statistics for all the months and all the columns. Let's try to retrieve only the mean, standard deviation and median values for all the months.

#### Activity 3: The `agg()` Function^^^

Instead of applying the aggregation functions on a `DataFrameGroupBy` object individually, you can apply them together using the `agg()` function. Inside the `agg()` function, you have to pass the names of the functions (as string values) as inputs.

**Syntax:** `DataFrameGroupBy.agg(func=('func1', 'func2' ... 'funcN'))`

where `'func1', 'func2' ... 'funcN'` are the names of the aggregation functions to be applied. They are passed as a singular value (through tuple) to the `func` parameter.

Let's apply the `mean(), std()` and `median()` function on the `group_2004_month DataFrameGroupBy` object.

In [ ]:
group_month_2004.agg(func=('mean','std','median'))

PT08.S1(CO)                       C6H6(GT)                   \
              mean         std  median       mean       std median   
Month                                                                
3      1222.819608  241.038541  1197.0   9.932745  7.060527   8.65   
4      1158.993056  227.826933  1115.5  10.329583  7.395352   8.80   
5      1076.350806  193.766317  1053.0  10.192339  6.670752   8.90   
6      1022.862500  157.922850  1030.5  10.500972  6.416231   9.00   
7      1046.452957  158.314118  1032.0  10.597043  6.227360   9.90   
8       986.259409  117.964334   987.5   7.112634  3.762033   6.60   
9      1081.022222  197.420922  1060.0  11.485694  7.918987   9.50   
10     1184.745968  243.017236  1159.5  13.518952  9.113308  11.10   
11     1132.148611  273.058383  1106.5  12.511806  9.838577  10.10   
12     1084.196237  211.089768  1060.0   9.423656  6.990846   8.60   

      PT08.S2(NMHC)                     PT08.S3(NOx)  ...     RH        AH  \
               mean         std  median         mean  ... median      mean   
Month                                                 ...                    
3        935.643137  258.997406   926.0  1029.172549  ...  52.20  0.789413   
4        948.565278  267.886642   931.0   931.405556  ...  50.10  0.927448   
5        950.735215  242.778219   937.0   945.151882  ...  44.30  0.959994   
6        964.565278  235.770258   939.0   896.379167  ...  41.40  1.266057   
7        971.427419  224.412396   972.5   804.911290  ...  33.10  1.242517   
8        844.643817  156.612666   840.0   835.005376  ...  45.50  1.585142   
9        991.970833  271.366865   958.5   810.725000  ...  44.75  1.280758   
10      1058.263441  294.244102  1019.0   688.321237  ...  63.00  1.463595   
11      1011.201389  327.342972   981.0   789.880556  ...  60.60  0.939991   
12       916.896505  253.792136   926.0   876.579301  ...  56.75  0.814761   

                          Year                    Day                   
            std   median  mean  std median       mean       std median  
Month                                                                   
3      0.139932  0.79205  2004  0.0   2004  20.870588  6.141338   21.0  
4      0.173082  0.92410  2004  0.0   2004  15.500000  8.661458   15.5  
5      0.199396  0.96495  2004  0.0   2004  16.000000  8.950289   16.0  
6      0.202981  1.25735  2004  0.0   2004  15.500000  8.661458   15.5  
7      0.243496  1.20525  2004  0.0   2004  16.000000  8.950289   16.0  
8      0.313015  1.64800  2004  0.0   2004  16.000000  8.950289   16.0  
9      0.293228  1.27195  2004  0.0   2004  15.500000  8.661458   15.5  
10     0.219919  1.51905  2004  0.0   2004  16.000000  8.950289   16.0  
11     0.360286  0.95040  2004  0.0   2004  15.500000  8.661458   15.5  
12     0.291196  0.85780  2004  0.0   2004  16.000000  8.950289   16.0  

[10 rows x 33 columns]

In [ ]:
group_month_2005.agg(func=('mean','std','median','min'))

PT08.S1(CO)                             C6H6(GT)                        \
              mean         std  median    min      mean       std median  min   
Month                                                                           
1      1108.086022  214.147270  1076.0  744.0  8.801210  6.948258    6.3  0.1   
2      1086.369048  184.077312  1070.0  755.0  7.365030  5.385039    6.3  0.2   
3      1145.223118  197.140374  1126.0  715.0  8.726478  6.677222    6.6  0.2   
4       953.712644  125.441352   925.0  811.0  4.285057  3.798394    3.3  0.8   

      PT08.S2(NMHC)              ...      AH          Year                    \
               mean         std  ...  median     min  mean  std median   min   
Month                            ...                                           
1        886.362903  266.509404  ...  0.6304  0.2477  2005  0.0   2005  2005   
2        840.401786  215.484273  ...  0.5208  0.2269  2005  0.0   2005  2005   
3        887.740591  253.315169  ...  0.7834  0.1847  2005  0.0   2005  2005   
4        698.735632  173.304307  ...  0.5977  0.3866  2005  0.0   2005  2005   

             Day                       
            mean       std median min  
Month                                  
1      16.000000  8.950289   16.0   1  
2      14.500000  8.083764   14.5   1  
3      16.000000  8.950289   16.0   1  
4       2.344828  1.065610    2.0   1  

[4 rows x 44 columns]

**Note:** You also choose to apply different aggregation functions on different columns by passing a dictionary containing the names of the columns and the aggregation function(s) to be applied on them.

E.g., let's apply the `mean(), std()` & `median()` functions on the `T` (temperature) column and `max()` function on the `AH` (Absolute humidity) column.

In [ ]:
group_month_2004.agg(func = {'T':['mean','std','median'],'AH':'max','Month':'count'})

T                       AH Month
            mean       std median     max count
Month                                          
3      14.390784  4.447539  14.05  1.0945   510
4      16.935694  4.758767  16.30  1.4852   720
5      20.283468  5.163229  19.30  1.6296   744
6      26.155833  5.955815  24.70  1.9390   720
7      29.406989  6.178302  28.05  2.0042   744
8      28.555376  5.469489  27.55  2.1806   744
9      24.514861  5.440697  23.60  2.2310   720
10     20.497446  3.586663  20.30  2.0224   744
11     13.485417  4.990447  13.25  1.9800   720
12     12.136425  4.462879  12.00  1.3713   744

#### Activity 4: Slicing `DataFrameGroupBy` Object^^


We still want to know the mean, standard deviation and median values only for the winter season for 2004 and 2005. 

So, we can use the `loc[]` function to get the rows only for the winter season.

In [ ]:
group_month_2004.agg(func = ['mean','std','median']).loc[[3,11,12],['AH','T']]

AH                             T                 
           mean       std   median       mean       std median
Month                                                         
3      0.789413  0.139932  0.79205  14.390784  4.447539  14.05
11     0.939991  0.360286  0.95040  13.485417  4.990447  13.25
12     0.814761  0.291196  0.85780  12.136425  4.462879  12.00

But we don't want any of the aggregated values for the `Year` and `Day` columns. Hence, we need to specify the column indices for the columns we want to be displayed.

In [ ]:
group_month_2004.agg(func=['mean','std','median']).columns

MultiIndex([(  'PT08.S1(CO)',   'mean'),
            (  'PT08.S1(CO)',    'std'),
            (  'PT08.S1(CO)', 'median'),
            (     'C6H6(GT)',   'mean'),
            (     'C6H6(GT)',    'std'),
            (     'C6H6(GT)', 'median'),
            ('PT08.S2(NMHC)',   'mean'),
            ('PT08.S2(NMHC)',    'std'),
            ('PT08.S2(NMHC)', 'median'),
            ( 'PT08.S3(NOx)',   'mean'),
            ( 'PT08.S3(NOx)',    'std'),
            ( 'PT08.S3(NOx)', 'median'),
            ( 'PT08.S4(NO2)',   'mean'),
            ( 'PT08.S4(NO2)',    'std'),
            ( 'PT08.S4(NO2)', 'median'),
            (  'PT08.S5(O3)',   'mean'),
            (  'PT08.S5(O3)',    'std'),
            (  'PT08.S5(O3)', 'median'),
            (            'T',   'mean'),
            (            'T',    'std'),
            (            'T', 'median'),
            (           'RH',   'mean'),
            (           'RH',    'std'),
            (           'RH', 'median'),
            (   

Here, we have a collection of column names. For each column, the column itself and the subcolumns (`mean, std, median`) are put together in a tuple. So, for every column, we have three tuples. 

We don't want the last six combinations of columns and sub columns to be part of our final aggregated DataFrame. Hence, we will pass the remaining column indices to the `loc[]` function.

In [ ]:
group_month_2004.agg(func=['mean','std','median']).loc[[3,11,12],group_month_2004.agg(func=['mean','std','median']).columns[:-6]]

PT08.S1(CO)                       C6H6(GT)                   \
              mean         std  median       mean       std median   
Month                                                                
3      1222.819608  241.038541  1197.0   9.932745  7.060527   8.65   
11     1132.148611  273.058383  1106.5  12.511806  9.838577  10.10   
12     1084.196237  211.089768  1060.0   9.423656  6.990846   8.60   

      PT08.S2(NMHC)                    PT08.S3(NOx)  ... PT08.S5(O3)  \
               mean         std median         mean  ...      median   
Month                                                ...               
3        935.643137  258.997406  926.0  1029.172549  ...       992.5   
11      1011.201389  327.342972  981.0   789.880556  ...      1150.5   
12       916.896505  253.792136  926.0   876.579301  ...       949.0   

               T                          RH                          AH  \
            mean       std median       mean        std median      mean   
Month                                                                      
3      14.390784  4.447539  14.05  50.170784  14.265312  52.20  0.789413   
11     13.485417  4.990447  13.25  59.274861  16.008765  60.60  0.939991   
12     12.136425  4.462879  12.00  56.825941  16.582928  56.75  0.814761   

                          
            std   median  
Month                     
3      0.139932  0.79205  
11     0.360286  0.95040  
12     0.291196  0.85780  

[3 rows x 27 columns]

The `group_2004_month.agg(func=['mean', 'std', 'median']).columns[:-6]` is a list of combinations of all the columns and sub columns except for the last 6 combinations.

In [ ]:
group_month_2005.agg(func=['mean','std','median']).loc[[1,2,3],group_month_2004.agg(func=['mean','std','median']).columns[:-6]]

PT08.S1(CO)                      C6H6(GT)                   \
              mean         std  median      mean       std median   
Month                                                               
1      1108.086022  214.147270  1076.0  8.801210  6.948258    6.3   
2      1086.369048  184.077312  1070.0  7.365030  5.385039    6.3   
3      1145.223118  197.140374  1126.0  8.726478  6.677222    6.6   

      PT08.S2(NMHC)                    PT08.S3(NOx)  ... PT08.S5(O3)  \
               mean         std median         mean  ...      median   
Month                                                ...               
1        886.362903  266.509404  826.0   788.237903  ...      1011.0   
2        840.401786  215.484273  826.0   786.099702  ...      1011.0   
3        887.740591  253.315169  841.0   709.447581  ...      1039.5   

               T                          RH                          AH  \
            mean       std median       mean        std median      mean   
Month                                                                      
1       8.294624  3.701922   8.10  56.208333  13.818780  54.75  0.634713   
2       7.338244  3.338817   7.35  52.096726  16.251039  52.20  0.530231   
3      13.213306  6.394258  13.50  50.909677  16.598389  51.80  0.781531   

                         
            std  median  
Month                    
1      0.218392  0.6304  
2      0.144846  0.5208  
3      0.291686  0.7834  

[3 rows x 27 columns]

**Note:** Unfortunately, the `quantile()` function cannot be applied as an aggregation function on a `DataFrameGroupBy` object. To get quartiles, use the `describe()` function.

Let's have a look on how we can get only quartiles using the `describe()` function.

In [ ]:
aq_2004_df.columns

Index(['DateTime', 'PT08.S1(CO)', 'C6H6(GT)', 'PT08.S2(NMHC)', 'PT08.S3(NOx)',
       'PT08.S4(NO2)', 'PT08.S5(O3)', 'T', 'RH', 'AH', 'Year', 'Month', 'Day',
       'Day Name'],
      dtype='object')

In [ ]:
for col in aq_2004_df.columns[1:4]:
    print(group_month_2004.describe().loc[[3,11,12],[(col,'25%'),(col,'50%'),(col,'75%')]],'\n')

      PT08.S1(CO)                
              25%     50%     75%
Month                            
3         1027.25  1197.0  1383.0
11         912.75  1106.5  1311.0
12         927.75  1060.0  1199.0 

      C6H6(GT)              
           25%    50%    75%
Month                       
3        4.600   8.65  13.65
11       5.000  10.10  17.60
12       4.375   8.60  12.50 

      PT08.S2(NMHC)                
                25%    50%      75%
Month                              
3            742.25  926.0  1106.25
11           762.50  981.0  1227.25
12           730.75  926.0  1067.00 



First, we have applied the `describe()` function which will return all the descriptive statistics for the `DataFrameGroupBy` object. Then using the `loc[]` function we have retrieved only the quartile values for the winter season of 2004.

In [ ]:
for col in aq_2005_df.columns[1:-4]:
    print(group_month_2005.describe().loc[[1,2,3],[(col,'25%'),(col,'50%'),(col,'75%')]],'\n')

      PT08.S1(CO)                 
              25%     50%      75%
Month                             
1          933.50  1076.0  1256.00
2          958.75  1070.0  1155.25
3          996.00  1126.0  1267.50 

      C6H6(GT)           
           25%  50%   75%
Month                    
1          3.6  6.3  13.1
2          3.6  6.3   9.5
3          3.7  6.6  12.1 

      PT08.S2(NMHC)                
                25%    50%      75%
Month                              
1            685.75  826.0  1087.25
2            688.75  826.0   960.00
3            692.50  841.0  1052.50 

      PT08.S3(NOx)               
               25%    50%     75%
Month                            
1           590.75  740.0  932.00
2           656.75  740.0  897.25
3           551.75  684.5  830.00 

      PT08.S4(NO2)                 
               25%     50%      75%
Month                              
1           948.00  1109.0  1319.75
2           917.00  1039.0  1160.25
3          1054.75  1217.0

**Note:** Since we have grouped the DataFrames for the years 2004 and 2005 by the `Month` column, it becomes the default index column in the aggregated DataFrames. If you don't want it to happen, pass the `as_index=False` parameter inside the `groupby()` function. 

In [ ]:
new_group_2004_month = aq_2004_df.groupby(by = 'Month',as_index = False)
new_group_2004_month.mean()

Month  PT08.S1(CO)   C6H6(GT)  PT08.S2(NMHC)  PT08.S3(NOx)  PT08.S4(NO2)  \
0      3  1222.819608   9.932745     935.643137   1029.172549   1572.072549   
1      4  1158.993056  10.329583     948.565278    931.405556   1608.054167   
2      5  1076.350806  10.192339     950.735215    945.151882   1599.932796   
3      6  1022.862500  10.500972     964.565278    896.379167   1712.698611   
4      7  1046.452957  10.597043     971.427419    804.911290   1643.682796   
5      8   986.259409   7.112634     844.643817    835.005376   1578.178763   
6      9  1081.022222  11.485694     991.970833    810.725000   1546.534722   
7     10  1184.745968  13.518952    1058.263441    688.321237   1636.818548   
8     11  1132.148611  12.511806    1011.201389    789.880556   1372.234722   
9     12  1084.196237   9.423656     916.896505    876.579301   1251.625000   

   PT08.S5(O3)          T         RH        AH    Year        Day  
0  1027.494118  14.390784  50.170784  0.789413  2004.0  20.870588  
1  1001.987500  16.935694  50.627569  0.927448  2004.0  15.500000  
2   928.598118  20.283468  43.636559  0.959994  2004.0  16.000000  
3   938.212500  26.155833  40.309861  1.266057  2004.0  15.500000  
4   994.391129  29.406989  33.075067  1.242517  2004.0  16.000000  
5   788.688172  28.555376  43.346035  1.585142  2004.0  16.000000  
6  1034.245833  24.514861  44.105833  1.280758  2004.0  15.500000  
7  1163.561828  20.497446  61.892944  1.463595  2004.0  16.000000  
8  1171.354167  13.485417  59.274861  0.939991  2004.0  15.500000  
9  1068.323925  12.136425  56.825941  0.814761  2004.0  16.000000

#### Activity 5: Multivariate Grouping & Aggregation

We can also group multiple columns at once by passing a list of columns to be grouped.

Let's group the DataFrames about the `Month` & `Day Name` columns.

In [ ]:
group_2004_month_day_name = aq_2004_df.groupby(by=['Month','Day Name'])


In [ ]:
group_2004_month_day_name.count()

DateTime  PT08.S1(CO)  C6H6(GT)  PT08.S2(NMHC)  PT08.S3(NOx)  \
Month Day Name                                                                  
3     Friday           72           72        72             72            72   
      Monday           72           72        72             72            72   
      Saturday         72           72        72             72            72   
      Sunday           72           72        72             72            72   
      Thursday         72           72        72             72            72   
...                   ...          ...       ...            ...           ...   
12    Saturday         96           96        96             96            96   
      Sunday           96           96        96             96            96   
      Thursday        120          120       120            120           120   
      Tuesday          96           96        96             96            96   
      Wednesday       120          120       120            120           120   

                 PT08.S4(NO2)  PT08.S5(O3)    T   RH   AH  Year  Day  
Month Day Name                                                        
3     Friday               72           72   72   72   72    72   72  
      Monday               72           72   72   72   72    72   72  
      Saturday             72           72   72   72   72    72   72  
      Sunday               72           72   72   72   72    72   72  
      Thursday             72           72   72   72   72    72   72  
...                       ...          ...  ...  ...  ...   ...  ...  
12    Saturday             96           96   96   96   96    96   96  
      Sunday               96           96   96   96   96    96   96  
      Thursday            120          120  120  120  120   120  120  
      Tuesday              96           96   96   96   96    96   96  
      Wednesday           120          120  120  120  120   120  120  

[70 rows x 12 columns]

Let's get all the records for all Mondays in the month of March, 2004. 

**Note:** The values to be retrieved should be passed as a tuple in the same order of grouping.

In [ ]:
group_2004_month_day_name.get_group((5,'Monday'))

DateTime  PT08.S1(CO)  C6H6(GT)  PT08.S2(NMHC)  PT08.S3(NOx)  \
1278 2004-05-03 00:00:00       1031.0       7.1          861.0         903.0   
1279 2004-05-03 01:00:00        917.0       3.0          654.0        1169.0   
1280 2004-05-03 02:00:00        888.0       2.2          598.0        1228.0   
1281 2004-05-03 03:00:00        872.0       1.7          564.0        1284.0   
1282 2004-05-03 04:00:00        876.0       1.5          547.0        1256.0   
...                  ...          ...       ...            ...           ...   
1969 2004-05-31 19:00:00       1124.0      16.6         1199.0         794.0   
1970 2004-05-31 20:00:00       1020.0      10.7         1003.0         904.0   
1971 2004-05-31 21:00:00       1035.0       7.5          878.0         977.0   
1972 2004-05-31 22:00:00        986.0       5.3          781.0        1063.0   
1973 2004-05-31 23:00:00       1017.0       6.5          834.0         986.0   

      PT08.S4(NO2)  PT08.S5(O3)     T    RH      AH  Year  Month  Day Day Name  
1278        1534.0        825.0  18.0  49.3  1.0086  2004      5    3   Monday  
1279        1414.0        545.0  17.6  54.6  1.0918  2004      5    3   Monday  
1280        1397.0        500.0  16.4  62.1  1.1493  2004      5    3   Monday  
1281        1385.0        494.0  15.6  65.3  1.1484  2004      5    3   Monday  
1282        1351.0        505.0  16.0  61.4  1.1043  2004      5    3   Monday  
...            ...          ...   ...   ...     ...   ...    ...  ...      ...  
1969        1858.0       1101.0  24.5  37.7  1.1454  2004      5   31   Monday  
1970        1716.0        916.0  21.7  46.9  1.2044  2004      5   31   Monday  
1971        1694.0        825.0  19.3  68.8  1.5190  2004      5   31   Monday  
1972        1633.0        693.0  17.6  81.6  1.6296  2004      5   31   Monday  
1973        1676.0        734.0  17.8  80.3  1.6212  2004      5   31   Monday  

[120 rows x 14 columns]

**Python Pandas - Merging/Joining**

Pandas has full-featured, high performance in-memory join operations idiomatically very similar to relational databases like SQL.

Pandas provides a single function, merge, as the entry point for all standard database join operations between DataFrame objects −

In [ ]:
left = pd.DataFrame({
    'id':[1,2,3,4,5],
    'Name':['Alex','as','dfg','df','fg'],
    'subid':['sub1','sub2','sub4','sub6','sub5']
})
right = pd.DataFrame({
    'id':[1,2,3,4,5],
    'Name':['a','b','c','d','f'],
    'subid':['sub2','sub4','sub3','sub6','sub5']
})

print(left)
right

   id  Name subid
0   1  Alex  sub1
1   2    as  sub2
2   3   dfg  sub4
3   4    df  sub6
4   5    fg  sub5


id Name subid
0   1    a  sub2
1   2    b  sub4
2   3    c  sub3
3   4    d  sub6
4   5    f  sub5

**Merge Two DataFrames on a Key**


In [ ]:
left = pd.DataFrame({
    'id':[1,2,3,4,5],
    'Name':['Alex','as','dfg','df','fg'],
    'subid':['sub1','sub2','sub4','sub6','sub5']
})
right = pd.DataFrame({
    'id':[1,2,3,4,5],
    'Name':['a','b','c','d','f'],
    'subid':['sub2','sub4','sub3','sub6','sub5']
})

print(pd.merge(left,right,on =['id']))

**Merge Two DataFrames on Multiple Keys**

In [ ]:
left = pd.DataFrame({
    'id':[1,2,3,4,5],
    'Name':['Alex','as','dfg','df','fg'],
    'subid':['sub1','sub2','sub4','sub6','sub5']
})
right = pd.DataFrame({
    'id':[1,2,3,4,5],
    'Name':['a','b','c','d','f'],
    'subid':['sub2','sub4','sub3','sub6','sub5']
})

print(pd.merge(left,right,on =['id','subid']))

   id Name_x subid Name_y
0   4     df  sub6      d
1   5     fg  sub5      f


**Python Pandas - Concatenation**

Concatenating Objects
The concat function does all of the heavy lifting of performing concatenation operations along an axis. Let us create different objects and do concatenation.

Suppose we wanted to associate specific keys with each of the pieces of the chopped up DataFrame. We can do this by using the keys argument −

The index of the resultant is duplicated; each index is repeated.

If the resultant object has to follow its own indexing, set ignore_index to True.

Observe, the index changes completely and the Keys are also overridden.

If two objects need to be added along axis=1, then the new columns will be appended.